### PR TITLE
Fixed cart not updated on voucher update

### LINKED ISSUE #19393: CART RULE - cart rule is applied even when you disabled
#### Describe the bug

If a Cart rule for cheapest product is created but not enabled and if a second cart rule is created and enabled the first one is displayed and enabled in the FO

#### Expected behavior

The cart rule must not be applied and displayed if not enabled

#### Steps to Reproduce

- [ ] Test1
1. Go to the BO => Catalog => Discounts => Create a new cart rule with the following parameters (see the attached screenshots below)
2. Enable the previously created cart rules
3. Go to the FO => Add to the cart demo_11 & demo_6
4. Go to the cart page => The cart rule is correctly applied to the cheapest product ( demo_11 )
5. Go Back to the BO => Catalog => Discounts => and disable the cart rule 
6. Go to the FO again and check the cart => the cart rule is applied and displayed => NOK

> ~7. Go Back to the BO => Catalog => Discounts => Create a new cart rule the parameters of this cart rule don't really matter just create a cart rule with a generated code and a amount of 10 % it's enough~
> ~8.Enable this cart rule~
> ~9. Go to the FO again and check the cart => the newly created cart rule is available~
> ~10. See error => the first created cart rule is applied and displayed even if not enabled in BO~

- [ ] Test2
>Edit by @hibatallahAouadni 

1. FO, add products to cart and apply a voucher
2. BO, edit the voucher date (valid to) to be expired
3. FO, refresh the page the voucher still applied => NOK
4. Delete the voucher, then you will see you can't apply it again => OK

- [ ] Test3
> Edited by Khouloud

Other steps to reproduce the same issue as described here https://github.com/PrestaShop/PrestaShop/issues/27650

**Preparation**
1. Create customer group ID 4.
2. Create a new customer, put him to group 4, set default group 4.
3. Create a cart rule that will apply automatically, cart rule will give a free gift, set group restriction only to 1, 2, 3.

**Actual bug**
1. Log in as the customer you created.
2. Add something to cart and go to cart.
3. Now, go edit that customer in BO, tick that he belongs to group 1.
4. Go back to cart, hit F5. Cart rule will be applied and gift added to cart.
5. Go back to edit customer, un-tick group 1, save.
6. Go back to cart, hit F5. Gift is not removed from cart.
_(7. It's removed when you +- the quantity on your product.)_
**Screen record**
https://drive.google.com/file/d/1SgVUMepMz-XsaUKgafB7aVv9LGnlAYLO/view

**Screenshots**

**Cart rule's parameters**
capture d'écran_35
capture d'écran_36
capture d'écran_37

**The cart rule is disabled in BO**
capture d'écran_43

**But applied and displayed in FO**
capture d'écran_44

#### Additional information

* PrestaShop version:1.7.7.x and 1.7.6.x
* PHP version: 7.2

### PR DESCRIPTION
**Pull request by SebBareyre**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Cart is not updated properly when there is a change in voucher 
| Type?             | bug fix
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #19393 (#20009 is a duplicate)
| How to test?      | Create a cart rule auto applicable (without code) and set the cart rule inactive or with an  expired date or without quantity available. The cart rule will be applicated. Try again with the patch, it will fix that.
| Possible impacts? | No impacts

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/23831)

### CODE REVIEW

**Comment by SebBareyre** [comment]
For information, you need to update the pull request template ;) 
the "critical" PR type was in the proposals...

File: classes/Cart.php
Comment (PierreRambaud) [inline classes/Cart.php:505]:
```suggestion
                AND NOW() BETWEEN cr.date_from AND cr.date_to
                AND cr.`active` = 1
                AND cr.`quantity` > 0
```

But I'm afraid you break integration tests, you change is breaking something :)

Comment (SebBareyre) [reply classes/Cart.php:505]: @Progi1984  Ok, I let you have a look on a manually test ;)

**Comment by Progi1984** [comment]
@SebBareyre Could you fix Behat issues ?

I think your PR has a bigger impact because the command `AddCartRuleToOrderCommand` (in `src/Core/Domain/Order/Command/AddCartRuleToOrderCommand.php` file) needs to be improved with a check if the cart rule is active, has quantity and is in date.

**Comment by SebBareyre** [comment]
@Progi1984 I let you do that 😉. I made the PR and I haven't more times to do that right now

**Comment by Progi1984** [comment]
@SebBareyre If #20009 is prioritzed by @PrestaShop/product-team, we work on it. It's why we are open to contributions and discussions to improve these contributions. If you don't really want, you can close your PR.

**Review by matks — CHANGES_REQUESTED** [review]
Hi, just like my maintainer peers said: the php class you are modifying is a critical piece of code, and modifying it could maybe solve one bug but introduce five others. This is why we need extra caution when we validate changes of this class, and the best extra caution we have as developers are tests.

Without tests, we cannot validate your changes in an automatic and long-term manner. Moreover the tests we already have are failing, which seems to indicate that indeed the changes of this PR introduce a buggy behavior.

Without tests acting like a safety net, I cannot approve this PR.

**Comment by PierreRambaud** [comment]
Hi @SebBareyre ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!

**Pull request closed without merging**
